application/src/main/java/org/togetherjava/tjbot/commands/code/CodeMessageHandler.java

```diff
@@ -0,0 +1,253 @@
+package org.togetherjava.tjbot.commands.code;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import net.dv8tion.jda.api.entities.Message;
+import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent;
+import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
+import net.dv8tion.jda.api.events.interaction.component.SelectMenuInteractionEvent;
+import net.dv8tion.jda.api.events.message.MessageDeleteEvent;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
+import net.dv8tion.jda.api.events.message.MessageUpdateEvent;
+import net.dv8tion.jda.api.interactions.components.buttons.Button;
+import net.dv8tion.jda.api.utils.messages.MessageCreateBuilder;
+import net.dv8tion.jda.api.utils.messages.MessageCreateData;
+import net.dv8tion.jda.internal.requests.CompletedRestAction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.togetherjava.tjbot.commands.MessageReceiverAdapter;
+import org.togetherjava.tjbot.commands.UserInteractionType;
+import org.togetherjava.tjbot.commands.UserInteractor;
+import org.togetherjava.tjbot.commands.componentids.ComponentIdGenerator;
+import org.togetherjava.tjbot.commands.componentids.ComponentIdInteractor;
+import org.togetherjava.tjbot.commands.utils.CodeFence;
+import org.togetherjava.tjbot.commands.utils.MessageUtils;
+
+import javax.annotation.Nullable;
+
+import java.awt.Color;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Handler that detects code in messages and offers code actions to the user, such as formatting
+ * their code.
+ * <p>
+ * Code actions are automatically updated whenever the code in the original message is edited or
+ * deleted.
+ */
+public final class CodeMessageHandler extends MessageReceiverAdapter implements UserInteractor {
+    private static final Logger logger = LoggerFactory.getLogger(CodeMessageHandler.class);
+
+    static final Color AMBIENT_COLOR = Color.decode("#FDFD96");
+
+    private final ComponentIdInteractor componentIdInteractor;
+    private final Map<String, CodeAction> labelToCodeAction;
+
+    /**
+     * Memorizes the ID of the bots code-reply message that a message belongs to. That way, the
+     * code-reply can be retrieved and managed easily when the original message is edited or
+     * deleted. Losing this cache, for example during bot-restart, effectively disables this
+     * update-feature for old messages.
+     * <p>
+     * The feature is secondary though, which is why its kept in RAM and not in the DB.
+     */
+    private final Cache<Long, Long> originalMessageToCodeReply =
+            Caffeine.newBuilder().maximumSize(10_000).build();
```

Comment:
@Tais993 10k should take almost no RAM at all, id estimate around 100 KB for a full cache. u need 1000 such caches to reach 1 GB RAM.

the point of the cache is exactly that, to prevent against RAM blowups. with a traditional `Map`, u have no proper limit and users are able to blow it up.

Reply:
done ✔️ 

reduced to 2k